File: src/main/java/br/com/jiratorio/domain/FluxColumn.java
package br.com.jiratorio.domain;

import br.com.jiratorio.domain.entity.Board;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;
import lombok.AllArgsConstructor;
import lombok.ToString;
import org.springframework.util.CollectionUtils;

@ToString
@AllArgsConstructor
public class FluxColumn {

    private String startLeadTimeColumn;

    private String endLeadTimeColumn;

    private List<String> orderedColumns;

    public FluxColumn(final Board board) {
        this(board.getStartColumn(), board.getEndColumn(), board.getFluxColumn());
    }

    public Set<String> getStartColumns() {
        Set<String> startColumns = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
        if (startLeadTimeColumn == null) {
            return startColumns;
        }

        startColumns.add(startLeadTimeColumn);
        if (!CollectionUtils.isEmpty(orderedColumns)) {
            int start = orderedColumns.indexOf(startLeadTimeColumn);
            int end = orderedColumns.indexOf(endLeadTimeColumn);
            if (start >= 0 && end >= 0 && start < end) {
                startColumns.addAll(orderedColumns.subList(start + 1, end + 1));
            }
        }
        return startColumns;
    }

    public Set<String> getEndColumns() {
        Set<String> endColumns = new TreeSet<>(String.CASE_INSENSITIVE_ORDER);
        if (endLeadTimeColumn == null) {
            return endColumns;
        }
        endColumns.add(endLeadTimeColumn);
        if (!CollectionUtils.isEmpty(orderedColumns)) {
            int start = orderedColumns.indexOf(endLeadTimeColumn);
            if (start >= 0 && start < orderedColumns.size() - 1) {
                endColumns.addAll(orderedColumns.subList(start + 1, orderedColumns.size()));
            }
        }
        return endColumns;
    }

    public Set<String> getWipColumns() {
        Set<String> wipColumns = getStartColumns();
        wipColumns.remove(endLeadTimeColumn);
        return wipColumns;
    }

    public String getLastColumn() {
        return CollectionUtils.isEmpty(orderedColumns) ? "Done" : orderedColumns.get(orderedColumns.size() - 1);
    }
}

File: src/main/java/br/com/jiratorio/domain/FluxColumn.kt
package br.com.jiratorio.domain

import br.com.jiratorio.domain.entity.Board
import lombok.AllArgsConstructor
import lombok.ToString

@ToString
@AllArgsConstructor
data class FluxColumn(
    val startLeadTimeColumn: String,
    val endLeadTimeColumn: String,
    val orderedColumns: List<String>?
) {

    constructor(board: Board) : this(board.startColumn!!, board.endColumn!!, board.fluxColumn)

    val endColumns: Set<String>
        get() = if (orderedColumns == null)
            setOf(endLeadTimeColumn)
        else
            orderedColumns.takeLastWhile { it != endLeadTimeColumn }.toSet() + endLeadTimeColumn

    val startColumns: Set<String>
        get() = if (orderedColumns == null)
            setOf(startLeadTimeColumn)
        else with(orderedColumns) {
            takeWhile { it != endLeadTimeColumn } - takeWhile { it != startLeadTimeColumn } + endLeadTimeColumn
        }.toSet()

    val wipColumns: Set<String?>
        get() = startColumns - endLeadTimeColumn

    val lastColumn: String
        get() = orderedColumns?.last() ?: "Done"

}
